user_sync/connector/directory_ldap.py

```diff
@@ -130,51 +146,70 @@ def load_users_and_groups(self, groups, extended_attributes, all_users):
         :rtype (bool, iterable(dict))
         """
         options = self.options
+        user = {}
+        base_dn = six.text_type(options['base_dn'])
         all_users_filter = six.text_type(options['all_users_filter'])
-        group_member_filter_format = six.text_type(options['group_member_filter_format'])
+        grouped_user_records = {}
+        if self.two_steps_lookup:
+            group_member_attribute_name = six.text_type(options['two_steps_lookup']['group_member_attribute_name'])
+
+        # save all the users to memory for faster 2-steps lookup or all_users process
+        if all_users:
+            try:
+                all_users_records = dict(self.iter_users(base_dn, all_users_filter, extended_attributes))
+            except Exception as e:
+                raise AssertionException('Unexpected LDAP failure reading all users: %s' % e)
 
         # for each group that's required, do one search for the users of that group
         for group in groups:
-            group_dn = self.find_ldap_group_dn(group)
+            group_users = 0
+            group_dn = self.find_group_dn(group)
             if not group_dn:
                 self.logger.warning("No group found for: %s", group)
                 continue
-            group_member_subfilter = self.format_ldap_query_string(group_member_filter_format, group_dn=group_dn)
-            if not group_member_subfilter.startswith('('):
-                group_member_subfilter = six.text_type('(') + group_member_subfilter + six.text_type(')')
-            user_subfilter = all_users_filter
-            if not user_subfilter.startswith('('):
-                user_subfilter = six.text_type('(') + user_subfilter + six.text_type(')')
-            group_user_filter = six.text_type('(&') + group_member_subfilter + user_subfilter + six.text_type(')')
-            group_users = 0
             try:
-                for user_dn, user in self.iter_users(group_user_filter, extended_attributes):
-                    user['groups'].append(group)
-                    group_users += 1
-                self.logger.debug('Count of users in group "%s": %d', group, group_users)
+                if self.two_steps_lookup:
+                    for user_dn in self.iter_group_member_dns(group_dn, group_member_attribute_name):
+                        #check to make sure user_dn are within the base_dn scope
+                        if self.is_dn_within_base_dn_scope(base_dn, user_dn):
+                            #replace base_dn with user_dn and filter with all_users_filter to do user lookup based on DN
+                            result =  list(self.iter_users(user_dn, all_users_filter, extended_attributes))
```

Comment:
PEP-8 nitpicks: Too many spaces after `=` operator, and there should be a space after the `#` on the two preceding comments.